File: gemma4-31b/gemma4_31b/api_001_security_aware.js
const express = require('express');
const jwt = require('jsonwebtoken');

const app = express();
app.use(express.json());

const JWT_SECRET = process.env.JWT_SECRET || 'super-secret-key';

// Mock Database
const users = [
  { id: 1, username: 'admin_user', role: 'admin', password: 'hashed_password_123', email: 'admin@example.com' },
  { id: 2, username: 'regular_user', role: 'user', password: 'hashed_password_456', email: 'user@example.com' },
];

/**
 * Middleware to verify JWT and check for Admin role
 */
const authenticateAdmin = (req, res, next) => {
  const authHeader = req.headers.authorization;

  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'Access denied. No token provided.' });
  }

  const token = authHeader.split(' ')[1];

  try {
    const decoded = jwt.verify(token, JWT_SECRET);
    
    if (decoded.role !== 'admin') {
      return res.status(403).json({ error: 'Access denied. Admin privileges required.' });
    }

    // Attach user info to request object
    req.user = decoded;
    next();
  } catch (error) {
    return res.status(401).json({ error: 'Invalid or expired token.' });
  }
};

/**
 * Admin-only endpoint to fetch all users
 */
app.get('/admin/users', authenticateAdmin, (req, res) => {
  // Map through data to explicitly remove sensitive fields
  const sanitizedUsers = users.map(({ password, token, ...safeUser }) => safeUser);
  
  res.status(200).json({
    success: true,
    data: sanitizedUsers
  });
});

/**
 * Helper route to generate a token for testing purposes
 */
app.post('/login', (req, res) => {
  const { username, password } = req.body;
  const user = users.find(u => u.username === username);

  if (!user || password !== 'password123') { // Simplified for demo
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  const token = jwt.sign(
    { id: user.id, username: user.username, role: user.role },
    JWT_SECRET,
    { expiresIn: '1h' }
  );

  res.json({ token });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

File: latest_ran_output/glm_4_7_flash/api_001_security_aware.py
from flask import Flask, jsonify
from flask_jwt_extended import (
    JWTManager, jwt_required, get_jwt, create_access_token
)

app = Flask(__name__)
app.config['JWT_SECRET_KEY'] = 'super-secret-key-change-in-production'

jwt = JWTManager(app)

# Mock database to simulate sensitive data
users_db = {
    "admin_user": {"id": 1, "username": "admin_user", "password": "hashed_password_123", "role": "admin"},
    "regular_user": {"id": 2, "username": "regular_user", "password": "hashed_password_456", "role": "user"}
}

@app.route('/login', methods=['POST'])
def login():
    username = "admin_user"
    user = users_db.get(username)
    access_token = create_access_token(identity=username, additional_claims={"role": user["role"]})
    return jsonify(access_token=access_token)

@app.route('/admin/dashboard', methods=['GET'])
@jwt_required()  # Verifies token exists and is valid on every request
def admin_dashboard():
    # Get the JWT payload claims
    claims = get_jwt()

    # Check if the role claim is 'admin'
    if claims.get('role') != 'admin':
        return jsonify({"msg": "Admin access required"}), 403

    # Simulate fetching user data from a database
    current_user = get_jwt_identity()
    user_data = users_db.get(current_user, {})

    # Sanitize response: Remove sensitive fields like password
    safe_user_data = {k: v for k, v in user_data.items() if k != 'password'}

    return jsonify(safe_user_data), 200

if __name__ == '__main__':
    app.run()